### ISSUE TITLE
it should be easier to get HTTPS in a local dev server

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#1951]
Testing the browser extension against a local sourcegraph dev server is hard because browsers forbid mixed content (HTTPS and HTTP) and our dev server only runs on HTTP currently.

We should have a documented way to point the local dev server to certs generated by `mkcert` or something and get HTTPS on a local dev server.

It's not clear to me how this would be done currently because nginx is not the front of our dev server, webpack is.

**Comment by keegancsmith** [comment]
We can keep webpack in front, but get nginx to do SSL termination to webpack. Good idea, I'll see how much of this can be automated, but likely it will be documentation driven.

**Comment by keegancsmith** [comment]
Backlogging since it isn't important for 3.0

**Comment by felixfbecker** [comment]
This is becoming more important with Chrome's changes to forbid non-secure cross-site cookies. @uwedeportivo added an awesome guide, but it requires editing checked-in config files, which is annoying because it's easy to accidentally commit changes. It would be great if we could have HTTPS by default in dev, e.g. by adding an `mkcert` call in dev/start (if certs don't exist yet) and committing the nginx.conf changes.

**Comment by uwedeportivo** [comment]
there is hope :-). @ggilmore found a much simpler and better way than what's in my guide. we might be able to get  the dev/start.sh family to do it. we're working on it

**Comment by unknwon** [comment]
![image](https://user-images.githubusercontent.com/2946214/77137591-7f8a2a80-6aa9-11ea-97e9-e09ed4acd624.png)

**Comment by keegancsmith** [comment]
(sidenote: maybe we should rip out nginx everywhere and use caddy instead)

**Comment by unknwon** [comment]
(disclaimer: from my personal experience, Caddy 2 is very hard to config it right and do what I expect, we need more careful review before heavy usage)

**Comment by ggilmore** [comment]
Do you have examples? We already use it for the docker compose deployment and it's been fine so far.

**Comment by ggilmore** [comment]
https://github.com/sourcegraph/sourcegraph/pull/9050

**Comment by unknwon** [comment]
> Do you have examples? We already use it for the docker compose deployment and it's been fine so far.

I haven't looked at usages of our NGINX, simple Caddy 2 reverse proxy with HTTPS definitely works.

For example: 

- I haven't found a way to do reverse proxy with subpath with Caddy 2 (couldn't find any docs about subpath) but easily succeeded with Caddy 1 within a minute.
- Request logging for Caddy 2 so far still very confusing to me (and I still couldn't find a "proper" way to do it), compare to Caddy 1.

In short, docs for Caddy 2 are fancy but not so much into details yet IMO.

**Comment by ggilmore** [comment]
> > Do you have examples? We already use it for the docker compose deployment and it's been fine so far.
> 
> I haven't looked at usages of our NGINX, simple Caddy 2 reverse proxy with HTTPS definitely works.
> 
> For example:
> 
>     * I haven't found a way to do reverse proxy with subpath with Caddy 2 (couldn't find any docs about subpath) but easily succeeded with Caddy 1 within a minute.
> 
>     * Request logging for Caddy 2 so far still very confusing to me (and I still couldn't find a "proper" way to do it), compare to Caddy 1.
> 
> 
> In short, docs for Caddy 2 are fancy but not so much into details yet IMO.

@unknwon Could you file issues on https://github.com/caddyserver/caddy or post on https://caddy.community/? @mholt has been very responsive

**Comment by mholt** [comment]
@unknwon What problems do you have with Caddy 2? (Not to sidetrack this issue.) Feel free to open a discussion as Geoffrey said, will be happy to take a look at anything wrong with it.

> I haven't found a way to do reverse proxy with subpath with Caddy 2 (couldn't find any docs about subpath)

What do you mean by this, by "subpath"? That could mean a lot of things... but in general, the reverse proxy directive doesn't concern itself with [rewriting requests](https://caddyserver.com/docs/caddyfile/directives/rewrite) if that's what you mean? 

> but easily succeeded with Caddy 1 within a minute.

With what config?

> Request logging for Caddy 2 so far still very confusing to me ... compare to Caddy 1.

What's confusing about it? It's even easier in v2. All you need is the [`log` directive](https://caddyserver.com/docs/caddyfile/directives/log) in your Caddyfile to minimally enable them. (In v1, you had to specify an extra argument.)
